---
title: "Code for simulation of cell type abundance analysis"
author: 
  name: Jun Inamo
  email: juninamo@keio.jp
  affiliation: Department of Microbiology and Immunology, Keio University School of Medicine, Tokyo, Japan
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  BiocStyle::html_document:
    toc_float: true

---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
# BiocManager::install("BiocStyle")
library(BiocStyle)

```

```{r, message=FALSE, warning=FALSE}

library(SimSCEQTL)

suppressWarnings({
  library(dplyr)
  library(tidyr)
  library(tidyverse)

  library(MASS)
  library(caret)
  
  library(Seurat)
  library(presto)
  library(ggplot2)
  library(ggrepel)
  library(glue)
  
  library(stevemisc)
  library(stevedata)
  library(lme4)
  library(broom.mixed)

  library(doParallel)
  library(pbapply)
  library(variancePartition)
  library(data.table)
  library(MatrixEQTL)
  library(harmony)
  library(foreach)
  library(parallel)

})

```

Simulation of cell type abundance analysis

# Scenario 1: without batch effect

```{r}
params <- list(
  n_thread = 14,
  n_cells = 100,
  n_individuals = 200,
  cluster_ratio = 0.65,
  batch_ratio = 0
)
```

```{r}
n_thread=params$n_thread
registerDoParallel(cores = n_thread)  # use forking with X cores

# parameters for generating dummy dataset
n_cells=params$n_cells # cells per major cell type
sd_celltypes=0.1  # standard deviation of number of cells 
n_major_cell_types=3 # number of major cell types
n_minor_cell_types=1 # number of minor cell types
relative_abundance=0.4 # ratio between major and rare
n_individuals=params$n_individuals # total individuals
n_batchs=4 # total batches
prop_sex=0.5 # proportion of female
prop_disease=0.5  # proportion of case
fc_increase=0.5 # additional FC of specified cell types which are from people with case groups compared to control groups (e.g., if you specify 0.5, FC comparing increased cell type between case and control groups will be 1.5 in total)
n_major_diff_celltypes=1 # number of major cell types which are differentially abundant between two groups. This parameter is effective only when fc_increase > 0
n_minor_diff_celltypes=1 # number of minor cell types which are differentially abundant between two groups. This parameter is effective only when fc_increase > 0

print(paste0("n_thread: ", n_thread))
print(paste0("n_cells: ", n_cells))
print(paste0("sd_celltypes: ", sd_celltypes))
print(paste0("n_major_cell_types: ", n_major_cell_types))
print(paste0("n_minor_cell_types: ", n_minor_cell_types))
print(paste0("relative_abundance: ", relative_abundance))
print(paste0("n_major_diff_celltypes: ", n_major_diff_celltypes))
print(paste0("n_minor_diff_celltypes: ", n_minor_diff_celltypes))
print(paste0("n_individuals: ", n_individuals))
print(paste0("n_batchs: ", n_batchs))
print(paste0("prop_sex: ", prop_sex))
print(paste0("prop_disease: ", prop_disease))
print(paste0("fc_increase: ", fc_increase))

n_features = 200
cluster_features = 1:50
disease_features = 1:50
sex_features = 1:50 # seq(2, 20, by = 2)
age_features = 1:50 # seq(2, 10, by = 2)
bmi_features = 1:50 # seq(11, 20, by = 2)
batch_features = 1:100 # seq(1, 20,by = 2)
individual_features = 1:100 # seq(1, 20,by = 2)

# Define the maximum variance in features of each attribute
cluster_ratio = params$cluster_ratio
disease_ratio = 0.01
sex_ratio = 0.01 # 0.05
age_ratio = 0.01 # 0.05
bmi_ratio = 0.01 # 0.05
batch_ratio = params$batch_ratio # 0.1
individual_ratio = 0.01 # 0.1

n_pcs = 20

ratio_variance = 0.1

cluster_col = "cell_type"
sex_col = "sex"
age_col = "age"
bmi_col = "bmi"
batch_col = "batch"
disease_col = "disease"
individual_col = 'subject_id'

# parameters for UMAP
n_neighbors = 30
metric = "cosine"
min_dist = 0.01

samplem_key = 'subject_id'

# parameters for eQTL
eqtl_celltype = "A" # major cell type
eqtl_gene = "Gene1" # gene to be tested for eQTL
cor_vals = c(0, 0.15, 0.3) # correlation coefficients to be tested between expression of eqtl_gene in eqtl_celltype and genotype (dosage)
allele_freqs = c(0.05, 0.3) # allele frequencies to be tested
n_snps = 100 # number of simulation SNPs
dist_type = "nb" # c("nb","poisson") # distribution of simulated counts. nb= negative binomial, poisson= Poisson distribution
seed = 1234
```

## Generating dummy metadata and feature matrix


```{r, warning=FALSE}
# Generate simulated data

datasets = generate_data(n_cells = n_cells,
                         sd_celltypes = sd_celltypes, 
                         n_major_cell_types = n_major_cell_types,
                         n_minor_cell_types = n_minor_cell_types,
                         relative_abundance = relative_abundance, 
                         n_major_diff_celltypes = n_major_diff_celltypes,
                         n_minor_diff_celltypes = n_minor_diff_celltypes,
                         
                         n_individuals = n_individuals,
                         n_batchs = n_batchs,
                         prop_sex = prop_sex, 
                         prop_disease = prop_disease, 
                         fc_increase = fc_increase, 
                         
                         n_features = n_features, 
                         cluster_features = cluster_features,
                         disease_features = disease_features,
                         sex_features = sex_features, 
                         age_features = age_features,
                         bmi_features = bmi_features,
                         batch_features = batch_features,
                         individual_features = individual_features,
                         
                         cluster_ratio = cluster_ratio,
                         disease_ratio = disease_ratio,
                         sex_ratio = sex_ratio,
                         age_ratio = age_ratio,
                         bmi_ratio = bmi_ratio,
                         batch_ratio = batch_ratio,
                         individual_ratio = individual_ratio,
                         
                         ratio_variance = ratio_variance,
                         
                         cluster_col = cluster_col,
                         sex_col = sex_col,
                         age_col = age_col,
                         bmi_col = bmi_col,
                         batch_col = batch_col,
                         disease_col = disease_col,
                         individual_col = individual_col,
                         
                         dist_type = dist_type,
                         eqtl_celltype = eqtl_celltype, 
                         eqtl_gene = eqtl_gene,
                         cor_vals = cor_vals,
                         allele_freqs = allele_freqs,
                         n_snps = n_snps, 
                         
                         seed = seed,
                         n_thread = n_thread,
                         verbose = TRUE
                         
                         )

```

```{r}
simulation_counts = datasets$simulation_counts
head(simulation_counts)
dim(simulation_counts)
meta_data = datasets$dummy_data
head(meta_data)
dim(meta_data)
```


## Inspect the generated data

```{r, fig.width=10, fig.height=6}

meta_data %>%
  dplyr::group_by(subject_id,sex,disease,cell_type) %>%
  dplyr::summarise(count = dplyr::n()) %>%
  dplyr::group_by(subject_id,sex,disease) %>%
  dplyr::mutate(proportion = 100*count/sum(count),
                sex_label = ifelse(sex=="0","Male","Female"),
                disease_label = ifelse(disease=="0","Control","Disease")) %>%
  ggplot(.,aes(x=disease_label,y=proportion,color = disease_label)) +
  geom_boxplot() +
  geom_point(position=position_jitterdodge(0.9)) +
  ggpubr::stat_compare_means(comparisons = list(c("Control","Disease")), 
                             label = "{p.adj}{p.adj.signif}", method = "wilcox.test", label.y.npc = 0.4, tip.length = 0) +
  scale_color_manual(values = c(
    "Control" = "#4DAF4A",
    "Disease" = "#984EA3")) +
  ggh4x::facet_nested_wrap(vars(cell_type, sex_label), nrow = 1) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15, angle = 45, hjust = 1),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)) +
  labs(
    x = "Disease status",
    y = "Proportion (%)",
    title = ""
  )  

```



## Normalization and PCA for simulated count data

```{r}

exprs_norm <- as(Matrix(as.matrix(simulation_counts), sparse = TRUE), "dgCMatrix") %>% NormalizeDataSeurat()
exprs_scaled <- t(scale(t(exprs_norm)))

exprs_cosine <- exprs_scaled %>% cosine_normalize(2)

# PCA for top 20 PCs
set.seed(1234)
pca_res <- irlba::prcomp_irlba(exprs_cosine, 20)
umap_res_fromCount <- uwot::umap(pca_res$x, n_neighbors = n_neighbors, metric = "cosine", min_dist = min_dist)
umap_res_fromCount = umap_res_fromCount %>% 
  magrittr::set_colnames(c("UMAP1","UMAP2")) %>%
  as.data.frame()

cluster_center <- data.frame(meta_data, umap_res_fromCount) %>%
  dplyr::group_by(cell_type) %>%
  dplyr::summarise_at(vars(UMAP1, UMAP2), dplyr::funs(median(., na.rm=TRUE)))
cluster_center <- as.data.frame(cluster_center)
ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "cell_type"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  geom_label_repel(
    data = cluster_center,
    aes(x = UMAP1, y = UMAP2, label = cell_type, fill = cell_type),
    color = "white",segment.color = "black",
    size = 7,
    min.segment.length = 0, seed = 42, box.padding = 0.8,
    max.overlaps = Inf
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = ""
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "subject_id"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by subject_id"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "disease"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by disease"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "batch"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by batch"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "age"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by age"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "bmi"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by bmi"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "sex"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by sex"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))
```



## Variance partitioning for simulated count data

```{r, fig.height=45, fig.width=6, warning=FALSE, message=FALSE}

set.seed(1234)

cl <- makeCluster(n_thread)

# Export the necessary objects and functions to the cluster
clusterExport(cl, list("apply_countmodel_to_column", "meta_data", "simulation_counts"))

# Load the necessary packages on each worker
clusterEvalQ(cl, {
  library(lme4)
  library(variancePartition)
})

results <- parLapply(cl, 1:ncol(simulation_counts), function(col_index) {
  column_data <- simulation_counts[, col_index]
  tryCatch({
    apply_countmodel_to_column(column_data, meta_data)
  }, error = function(e) {
    vars_na = rep(NA,8)
    names(vars_na) = c('batch', 'cell_type', 'disease', 'sex', 'subject_id', 'age', 'bmi', 'Residuals')
    return(vars_na)
    # return(list(error = e$message))
  })
})
stopCluster(cl)


# Assuming 'results' is your list of variance partitions
# Convert the list of results to a data frame
results_df <- do.call(rbind, lapply(results, function(x) as.data.frame(t(x))))
colnames(results_df) <- c('batch', 'cell_type', 'disease', 'sex', 'subject_id', 'age', 'bmi', 'Residuals')
# results_df = na.omit(results_df)
results_df$Gene <- paste0("Gene", 1:ncol(simulation_counts))

results_long <- tidyr::gather(results_df, key = "variable", value = "variance", -Gene)

# Compute the percentage of variance explained by each variable for each Gene
results_long <- results_long %>%
  dplyr::group_by(Gene) %>%
  dplyr::mutate(percentage = (variance / sum(variance)) * 100) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(#Gene = factor(Gene, levels = paste0("Gene", c(1:10,(ncol(simulation_counts)-9):ncol(simulation_counts)))),
    Gene = factor(Gene, levels = paste0("Gene", 1:ncol(simulation_counts))),
    variable = factor(variable, levels = c('cell_type', 'batch', 'disease', 'sex', 'subject_id', 'age', 'bmi', 'Residuals')))

# Create the stacked bar plot
ggplot(na.omit(results_long), aes(x = Gene, y = percentage, fill = variable)) +
  geom_bar(stat = "identity") +
  labs(x = "Genes", y = "Variance Explained (%)", title = "Variance Explained") +
  coord_flip() +
  scale_fill_brewer(palette = "Dark2") +
  theme_minimal() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)
  )

```


```{r, fig.height=5, fig.width=6, warning=FALSE}

set.seed(1234)
# Set up a cluster using the available cores minus one
cl <- makeCluster(n_thread)

pcs_matrix = pca_res$x
# Export the necessary objects and functions to the cluster
clusterExport(cl, list("apply_model_to_column", "meta_data", "pcs_matrix"))

# Load the necessary packages on each worker
clusterEvalQ(cl, {
  library(lme4)
  library(variancePartition)
})

# Use parLapply to apply the function in parallel and capture errors
results <- parLapply(cl, 1:ncol(pcs_matrix), function(col_index) {
  column_data <- pcs_matrix[, col_index]
  tryCatch({
    apply_model_to_column(column_data, meta_data)
  }, error = function(e) {
    return(list(error = e$message))
  })
})

stopCluster(cl)

# Check for errors
## errors <- sapply(results, function(x) if(is.list(x)) x$error else NA)
## errors

# Assuming 'results' is your list of variance partitions
# Convert the list of results to a data frame
results = results[!sapply(results, function(x) is.list(x) && !is.null(x$error))] # remove errors
results_df <- do.call(rbind, lapply(results, function(x) as.data.frame(t(x))))
colnames(results_df) <- c('batch', 'cell_type', 'disease', 'sex', 'subject_id', 'age', 'bmi', 'Residuals')
results_df$PC <- paste0("PC", rownames(results_df))

# Convert the results into a long format for ggplot
results_long <- tidyr::gather(results_df, key = "variable", value = "variance", -PC)

# Compute the percentage of variance explained by each variable for each PC
results_long <- results_long %>%
  dplyr::group_by(PC) %>%
  dplyr::mutate(percentage = (variance / sum(variance)) * 100) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(PC = factor(PC, levels = paste0("PC", 1:ncol(pcs_matrix))),
                variable = factor(variable, levels = c('cell_type', 'batch', 'disease', 'sex', 'subject_id', 'age', 'bmi', 'Residuals')))

# Create the stacked bar plot
ggplot(results_long, aes(x = PC, y = percentage, fill = variable)) +
  geom_bar(stat = "identity") +
  labs(x = "Principal Component", y = "Variance Explained (%)", title = "Variance Explained") +
  coord_flip() +
  scale_fill_brewer(palette = "Dark2") +
  theme_minimal() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)
        )


```


```{r, message=FALSE, warning=FALSE, fig.height=6, fig.width=12}
meta_data_with_pcs <- cbind(meta_data, pcs_matrix)

meta_data_with_pcs %>%
  dplyr::select(cell_type, starts_with("PC")) %>%
  pivot_longer(!cell_type, names_to = "PC", values_to = "value") %>%
  dplyr::mutate(PC = factor(PC, levels=paste0("PC",1:20))) %>%
  ggplot(., aes(x = value, y = cell_type)) +
  ggridges::geom_density_ridges(aes(fill = cell_type, color = cell_type), scale = 1) +
  facet_wrap(PC ~ ., ncol = 10, scales = "free") +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)) +
  labs(title = "",
       x = "loadings",
       y = "cell clusters") 
meta_data_with_pcs %>%
  dplyr::select(disease, starts_with("PC")) %>%
  pivot_longer(!disease, names_to = "PC", values_to = "value") %>%
  dplyr::mutate(PC = factor(PC, levels=paste0("PC",1:20))) %>%
  ggplot(., aes(x = value, y = disease)) +
  ggridges::geom_density_ridges(aes(fill = disease, color = disease), scale = 1) +
  facet_wrap(PC ~ ., ncol = 10, scales = "free") +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)) +
  labs(title = "",
       x = "loadings",
       y = "disease") 
meta_data_with_pcs %>%
  dplyr::select(batch, starts_with("PC")) %>%
  pivot_longer(!batch, names_to = "PC", values_to = "value") %>%
  dplyr::mutate(PC = factor(PC, levels=paste0("PC",1:20))) %>%
  ggplot(., aes(x = value, y = batch)) +
  ggridges::geom_density_ridges(aes(fill = batch, color = batch), scale = 1) +
  facet_wrap(PC ~ ., ncol = 10, scales = "free") +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)) +
  labs(title = "",
       x = "loadings",
       y = "batch") 
meta_data_with_pcs %>%
  dplyr::select(sex, starts_with("PC")) %>%
  pivot_longer(!sex, names_to = "PC", values_to = "value") %>%
  dplyr::mutate(PC = factor(PC, levels=paste0("PC",1:20))) %>%
  ggplot(., aes(x = value, y = sex)) +
  ggridges::geom_density_ridges(aes(fill = sex, color = sex), scale = 1) +
  facet_wrap(PC ~ ., ncol = 10, scales = "free") +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)) +
  labs(title = "",
       x = "loadings",
       y = "sex") 

```


# Scenario 2: with batch effect

```{r}
params <- list(
  n_thread = 14,
  n_cells = 100,
  n_individuals = 200,
  cluster_ratio = 0.65,
  batch_ratio = 0.3
)
```

```{r}
n_thread=params$n_thread
registerDoParallel(cores = n_thread)  # use forking with X cores

# parameters for generating dummy dataset
n_cells=params$n_cells # cells per major cell type
sd_celltypes=0.1  # standard deviation of number of cells 
n_major_cell_types=3 # number of major cell types
n_minor_cell_types=1 # number of minor cell types
relative_abundance=0.4 # ratio between major and rare
n_individuals=params$n_individuals # total individuals
n_batchs=4 # total batches
prop_sex=0.5 # proportion of female
prop_disease=0.5  # proportion of case
fc_increase=0.5 # additional FC of specified cell types which are from people with case groups compared to control groups (e.g., if you specify 0.5, FC comparing increased cell type between case and control groups will be 1.5 in total)
n_major_diff_celltypes=1 # number of major cell types which are differentially abundant between two groups. This parameter is effective only when fc_increase > 0
n_minor_diff_celltypes=1 # number of minor cell types which are differentially abundant between two groups. This parameter is effective only when fc_increase > 0

print(paste0("n_thread: ", n_thread))
print(paste0("n_cells: ", n_cells))
print(paste0("sd_celltypes: ", sd_celltypes))
print(paste0("n_major_cell_types: ", n_major_cell_types))
print(paste0("n_minor_cell_types: ", n_minor_cell_types))
print(paste0("relative_abundance: ", relative_abundance))
print(paste0("n_major_diff_celltypes: ", n_major_diff_celltypes))
print(paste0("n_minor_diff_celltypes: ", n_minor_diff_celltypes))
print(paste0("n_individuals: ", n_individuals))
print(paste0("n_batchs: ", n_batchs))
print(paste0("prop_sex: ", prop_sex))
print(paste0("prop_disease: ", prop_disease))
print(paste0("fc_increase: ", fc_increase))

n_features = 200
cluster_features = 1:50
disease_features = 1:50
sex_features = 1:50 # seq(2, 20, by = 2)
age_features = 1:50 # seq(2, 10, by = 2)
bmi_features = 1:50 # seq(11, 20, by = 2)
batch_features = 1:100 # seq(1, 20,by = 2)
individual_features = 1:100 # seq(1, 20,by = 2)

# Define the maximum variance in features of each attribute
cluster_ratio = params$cluster_ratio
disease_ratio = 0.01
sex_ratio = 0.01 # 0.05
age_ratio = 0.01 # 0.05
bmi_ratio = 0.01 # 0.05
batch_ratio = params$batch_ratio # 0.1
individual_ratio = 0.01 # 0.1

n_pcs = 20

ratio_variance = 0.1

cluster_col = "cell_type"
sex_col = "sex"
age_col = "age"
bmi_col = "bmi"
batch_col = "batch"
disease_col = "disease"
individual_col = 'subject_id'

# parameters for UMAP
n_neighbors = 30
metric = "cosine"
min_dist = 0.01

samplem_key = 'subject_id'

# parameters for eQTL
eqtl_celltype = "A" # major cell type
eqtl_gene = "Gene1" # gene to be tested for eQTL
cor_vals = c(0, 0.15, 0.3) # correlation coefficients to be tested between expression of eqtl_gene in eqtl_celltype and genotype (dosage)
allele_freqs = c(0.05, 0.3) # allele frequencies to be tested
n_snps = 100 # number of simulation SNPs
dist_type = "nb" # c("nb","poisson") # distribution of simulated counts. nb= negative binomial, poisson= Poisson distribution
seed = 1234
```

## Generating dummy metadata and feature matrix


```{r, warning=FALSE}
# Generate simulated data

datasets = generate_data(n_cells = n_cells,
                         sd_celltypes = sd_celltypes, 
                         n_major_cell_types = n_major_cell_types,
                         n_minor_cell_types = n_minor_cell_types,
                         relative_abundance = relative_abundance, 
                         n_major_diff_celltypes = n_major_diff_celltypes,
                         n_minor_diff_celltypes = n_minor_diff_celltypes,
                         
                         n_individuals = n_individuals,
                         n_batchs = n_batchs,
                         prop_sex = prop_sex, 
                         prop_disease = prop_disease, 
                         fc_increase = fc_increase, 
                         
                         n_features = n_features, 
                         cluster_features = cluster_features,
                         disease_features = disease_features,
                         sex_features = sex_features, 
                         age_features = age_features,
                         bmi_features = bmi_features,
                         batch_features = batch_features,
                         individual_features = individual_features,
                         
                         cluster_ratio = cluster_ratio,
                         disease_ratio = disease_ratio,
                         sex_ratio = sex_ratio,
                         age_ratio = age_ratio,
                         bmi_ratio = bmi_ratio,
                         batch_ratio = batch_ratio,
                         individual_ratio = individual_ratio,
                         
                         ratio_variance = ratio_variance,
                         
                         cluster_col = cluster_col,
                         sex_col = sex_col,
                         age_col = age_col,
                         bmi_col = bmi_col,
                         batch_col = batch_col,
                         disease_col = disease_col,
                         individual_col = individual_col,
                         
                         dist_type = dist_type,
                         eqtl_celltype = eqtl_celltype, 
                         eqtl_gene = eqtl_gene,
                         cor_vals = cor_vals,
                         allele_freqs = allele_freqs,
                         n_snps = n_snps, 
                         
                         seed = seed,
                         n_thread = n_thread,
                         verbose = TRUE
                         
                         )

```

```{r}
simulation_counts = datasets$simulation_counts
head(simulation_counts)
dim(simulation_counts)
meta_data = datasets$dummy_data
head(meta_data)
dim(meta_data)
```

## Inspect the generated data

```{r, fig.width=10, fig.height=6}

meta_data %>%
  dplyr::group_by(subject_id,sex,disease,cell_type) %>%
  dplyr::summarise(count = dplyr::n()) %>%
  dplyr::group_by(subject_id,sex,disease) %>%
  dplyr::mutate(proportion = 100*count/sum(count),
                sex_label = ifelse(sex=="0","Male","Female"),
                disease_label = ifelse(disease=="0","Control","Disease")) %>%
  ggplot(.,aes(x=disease_label,y=proportion,color = disease_label)) +
  geom_boxplot() +
  geom_point(position=position_jitterdodge(0.9)) +
  ggpubr::stat_compare_means(comparisons = list(c("Control","Disease")), 
                             label = "{p.adj}{p.adj.signif}", method = "wilcox.test", label.y.npc = 0.4, tip.length = 0) +
  scale_color_manual(values = c(
    "Control" = "#4DAF4A",
    "Disease" = "#984EA3")) +
  ggh4x::facet_nested_wrap(vars(cell_type, sex_label), nrow = 1) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15, angle = 45, hjust = 1),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)) +
  labs(
    x = "Disease status",
    y = "Proportion (%)",
    title = ""
  )  

```



## Normalization and PCA for simulated count data

```{r}

exprs_norm <- as(Matrix(as.matrix(simulation_counts), sparse = TRUE), "dgCMatrix") %>% NormalizeDataSeurat()
exprs_scaled <- t(scale(t(exprs_norm)))

exprs_cosine <- exprs_scaled %>% cosine_normalize(2)

# PCA for top 20 PCs
set.seed(1234)
pca_res <- irlba::prcomp_irlba(exprs_cosine, 20)
umap_res_fromCount <- uwot::umap(pca_res$x, n_neighbors = n_neighbors, metric = "cosine", min_dist = min_dist)
umap_res_fromCount = umap_res_fromCount %>% 
  magrittr::set_colnames(c("UMAP1","UMAP2")) %>%
  as.data.frame()

cluster_center <- data.frame(meta_data, umap_res_fromCount) %>%
  dplyr::group_by(cell_type) %>%
  dplyr::summarise_at(vars(UMAP1, UMAP2), dplyr::funs(median(., na.rm=TRUE)))
cluster_center <- as.data.frame(cluster_center)
ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "cell_type"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  geom_label_repel(
    data = cluster_center,
    aes(x = UMAP1, y = UMAP2, label = cell_type, fill = cell_type),
    color = "white",segment.color = "black",
    size = 7,
    min.segment.length = 0, seed = 42, box.padding = 0.8,
    max.overlaps = Inf
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = ""
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "subject_id"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by subject_id"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "disease"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by disease"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "batch"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by batch"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "age"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by age"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "bmi"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by bmi"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))

ggplot() +
  geom_point(
    data = data.frame(meta_data, umap_res_fromCount) %>% sample_n(nrow(.)),
    mapping = aes_string(x = "UMAP1", y = "UMAP2", color = "sex"),
    size = 1, stroke = 0, shape = 16,alpha = 0.3
  ) +
  labs(
    x = "UMAP1",
    y = "UMAP2",
    title = "color by sex"
  ) +
  coord_fixed(ratio = 1.2) +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black"),
        strip.text.y=element_text(size=15, color="black"),
        strip.placement = "outside", 
        strip.background.x=element_rect(color = NA,  fill=NA), 
        strip.background.y=element_rect(color = "black",  fill=NA),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 10),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 10, hjust = 0))
```



## Variance partitioning for simulated count data

```{r, fig.height=45, fig.width=6, warning=FALSE, message=FALSE}

set.seed(1234)

cl <- makeCluster(n_thread)

# Export the necessary objects and functions to the cluster
clusterExport(cl, list("apply_countmodel_to_column", "meta_data", "simulation_counts"))

# Load the necessary packages on each worker
clusterEvalQ(cl, {
  library(lme4)
  library(variancePartition)
})

results <- parLapply(cl, 1:ncol(simulation_counts), function(col_index) {
  column_data <- simulation_counts[, col_index]
  tryCatch({
    apply_countmodel_to_column(column_data, meta_data)
  }, error = function(e) {
    vars_na = rep(NA,8)
    names(vars_na) = c('batch', 'cell_type', 'disease', 'sex', 'subject_id', 'age', 'bmi', 'Residuals')
    return(vars_na)
    # return(list(error = e$message))
  })
})
stopCluster(cl)


# Assuming 'results' is your list of variance partitions
# Convert the list of results to a data frame
results_df <- do.call(rbind, lapply(results, function(x) as.data.frame(t(x))))
colnames(results_df) <- c('batch', 'cell_type', 'disease', 'sex', 'subject_id', 'age', 'bmi', 'Residuals')
# results_df = na.omit(results_df)
results_df$Gene <- paste0("Gene", 1:ncol(simulation_counts))

results_long <- tidyr::gather(results_df, key = "variable", value = "variance", -Gene)

# Compute the percentage of variance explained by each variable for each Gene
results_long <- results_long %>%
  dplyr::group_by(Gene) %>%
  dplyr::mutate(percentage = (variance / sum(variance)) * 100) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(#Gene = factor(Gene, levels = paste0("Gene", c(1:10,(ncol(simulation_counts)-9):ncol(simulation_counts)))),
    Gene = factor(Gene, levels = paste0("Gene", 1:ncol(simulation_counts))),
    variable = factor(variable, levels = c('cell_type', 'batch', 'disease', 'sex', 'subject_id', 'age', 'bmi', 'Residuals')))

# Create the stacked bar plot
ggplot(na.omit(results_long), aes(x = Gene, y = percentage, fill = variable)) +
  geom_bar(stat = "identity") +
  labs(x = "Genes", y = "Variance Explained (%)", title = "Variance Explained") +
  coord_flip() +
  scale_fill_brewer(palette = "Dark2") +
  theme_minimal() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)
  )

```


```{r, fig.height=5, fig.width=6, warning=FALSE}

set.seed(1234)

# Set up a cluster using the available cores minus one
cl <- makeCluster(n_thread)

pcs_matrix = pca_res$x
# Export the necessary objects and functions to the cluster
clusterExport(cl, list("apply_model_to_column", "meta_data", "pcs_matrix"))

# Load the necessary packages on each worker
clusterEvalQ(cl, {
  library(lme4)
  library(variancePartition)
})

# Use parLapply to apply the function in parallel and capture errors
results <- parLapply(cl, 1:ncol(pcs_matrix), function(col_index) {
  column_data <- pcs_matrix[, col_index]
  tryCatch({
    apply_model_to_column(column_data, meta_data)
  }, error = function(e) {
    return(list(error = e$message))
  })
})

stopCluster(cl)

# Check for errors
## errors <- sapply(results, function(x) if(is.list(x)) x$error else NA)
## errors

# Assuming 'results' is your list of variance partitions
# Convert the list of results to a data frame
results = results[!sapply(results, function(x) is.list(x) && !is.null(x$error))] # remove errors
results_df <- do.call(rbind, lapply(results, function(x) as.data.frame(t(x))))
colnames(results_df) <- c('batch', 'cell_type', 'disease', 'sex', 'subject_id', 'age', 'bmi', 'Residuals')
results_df$PC <- paste0("PC", rownames(results_df))

# Convert the results into a long format for ggplot
results_long <- tidyr::gather(results_df, key = "variable", value = "variance", -PC)

# Compute the percentage of variance explained by each variable for each PC
results_long <- results_long %>%
  dplyr::group_by(PC) %>%
  dplyr::mutate(percentage = (variance / sum(variance)) * 100) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(PC = factor(PC, levels = paste0("PC", 1:ncol(pcs_matrix))),
                variable = factor(variable, levels = c('cell_type', 'batch', 'disease', 'sex', 'subject_id', 'age', 'bmi', 'Residuals')))

# Create the stacked bar plot
ggplot(results_long, aes(x = PC, y = percentage, fill = variable)) +
  geom_bar(stat = "identity") +
  labs(x = "Principal Component", y = "Variance Explained (%)", title = "Variance Explained") +
  coord_flip() +
  scale_fill_brewer(palette = "Dark2") +
  theme_minimal() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "right",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)
        )


```


```{r, message=FALSE, warning=FALSE, fig.height=6, fig.width=12}
meta_data_with_pcs <- cbind(meta_data, pcs_matrix)

meta_data_with_pcs %>%
  dplyr::select(cell_type, starts_with("PC")) %>%
  pivot_longer(!cell_type, names_to = "PC", values_to = "value") %>%
  dplyr::mutate(PC = factor(PC, levels=paste0("PC",1:20))) %>%
  ggplot(., aes(x = value, y = cell_type)) +
  ggridges::geom_density_ridges(aes(fill = cell_type, color = cell_type), scale = 1) +
  facet_wrap(PC ~ ., ncol = 10, scales = "free") +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)) +
  labs(title = "",
       x = "loadings",
       y = "cell clusters") 
meta_data_with_pcs %>%
  dplyr::select(disease, starts_with("PC")) %>%
  pivot_longer(!disease, names_to = "PC", values_to = "value") %>%
  dplyr::mutate(PC = factor(PC, levels=paste0("PC",1:20))) %>%
  ggplot(., aes(x = value, y = disease)) +
  ggridges::geom_density_ridges(aes(fill = disease, color = disease), scale = 1) +
  facet_wrap(PC ~ ., ncol = 10, scales = "free") +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)) +
  labs(title = "",
       x = "loadings",
       y = "disease") 
meta_data_with_pcs %>%
  dplyr::select(batch, starts_with("PC")) %>%
  pivot_longer(!batch, names_to = "PC", values_to = "value") %>%
  dplyr::mutate(PC = factor(PC, levels=paste0("PC",1:20))) %>%
  ggplot(., aes(x = value, y = batch)) +
  ggridges::geom_density_ridges(aes(fill = batch, color = batch), scale = 1) +
  facet_wrap(PC ~ ., ncol = 10, scales = "free") +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)) +
  labs(title = "",
       x = "loadings",
       y = "batch") 
meta_data_with_pcs %>%
  dplyr::select(sex, starts_with("PC")) %>%
  pivot_longer(!sex, names_to = "PC", values_to = "value") %>%
  dplyr::mutate(PC = factor(PC, levels=paste0("PC",1:20))) %>%
  ggplot(., aes(x = value, y = sex)) +
  ggridges::geom_density_ridges(aes(fill = sex, color = sex), scale = 1) +
  facet_wrap(PC ~ ., ncol = 10, scales = "free") +
  theme_classic() +
  theme(strip.text.x=element_text(size=15, color="black", face="bold"),
        strip.text.y=element_text(size=15, color="black", face="bold"),
        legend.position = "none",
        plot.title = element_text(size=15),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size =15),
        axis.text.y = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.text =  element_text(size = 15),
        legend.key.size = grid::unit(0.5, "lines"),
        legend.title = element_text(size = 0.8, hjust = 0)) +
  labs(title = "",
       x = "loadings",
       y = "sex") 

```
